File: entity-meta-generator-lib/src/main/java/com/loudsight/meta/serialization/TypeConverters.java
package com.loudsight.meta.serialization;

import com.loudsight.useful.helper.ClassHelper;
import com.loudsight.useful.helper.JvmClassHelper;

import java.time.LocalDateTime;
import java.time.ZonedDateTime;
import java.util.HashMap;
import java.util.Map;

public final class TypeConverters {
    private static final class TypeConvertersHolder {
        private static final TypeConverters INSTANCE = new TypeConverters();
    }
    // global access point
    public static TypeConverters getInstance() {
        return TypeConverters.TypeConvertersHolder.INSTANCE;
    }
    private static final Map<Class<?>, Map<Class<Object>, TypeConverter<?, ?>>> converters = new HashMap<>();

    private final Map<Class<Object>, TypeConverter<?, ?>> NO_CONVERTERS = new HashMap<>();

    private TypeConverters() {
        register(Integer.class, Long.class, from -> from != null ? from.intValue() : null);
        register(Class.class, String.class, JvmClassHelper::classForName);
        register(LocalDateTime.class, ZonedDateTime.class, LocalDateTime::from);
    }

    <T, F> void register(Class<T> to, Class<F> from, TypeConverter<T, F> converter) {
        var typeConverters = converters.compute(to, (k, v) -> new HashMap<>());
        typeConverters.put(ClassHelper.uncheckedCast(from), converter);
    }

    public  <T> T convert(Object from, Class<T> toType) {
        if (toType == from.getClass()) {
            return ClassHelper.uncheckedCast(from);
        }
        var fromConverter = converters.getOrDefault(toType, NO_CONVERTERS);
        var fromKClass = from.getClass();
        var toConverter = fromConverter.getOrDefault(fromKClass, new NoOpConverter());

        return ClassHelper.uncheckedCast(toConverter.convert(ClassHelper.uncheckedCast(from)));
    }

    static class NoOpConverter implements TypeConverter<Object, Object> {
        @Override
        public Object convert(Object from) {
            return from;
        }
    }
}

File: entity-meta-generator-lib/src/main/kotlin/com/loudsight/meta/serialization/TypeConverters.kt
package com.loudsight.meta.serialization

import com.loudsight.useful.helper.ClassHelper
import kotlin.reflect.KClass

object TypeConverters {
    private val converters = mutableMapOf<KClass<Any>, MutableMap<KClass<Any>, TypeConverter<*, *>>>()

    private val NO_CONVERTERS: MutableMap<KClass<Any>, TypeConverter<*, *>> = mutableMapOf()

    init {
        register(Int::class, Long::class, object : TypeConverter<Int, Long> {
            override fun convert(from: Long?): Int {
                return from!!.toInt()
            }
        })

    }

    fun <T : Any, F : Any> register(to: KClass<T>, from: KClass<F>, converter: TypeConverter<T, F>) {
        val typeConverters = converters.getOrPut(to as KClass<Any>) { mutableMapOf() }
        typeConverters[from as KClass<Any>] = converter
    }

    fun <T : Any> convert(from: Any, toType: KClass<T>): T? {
        if (toType == from::class) {
            return from as T
        }
        val fromConverter = converters.getOrElse(toType as KClass<Any>) { NO_CONVERTERS }
        val fromKClass = ClassHelper.uncheckedCast<KClass<Any>>(from::class)
        val toConverter = fromConverter.getOrElse(fromKClass) { NoOpConverter() } as TypeConverter<Any, Any>

        return toConverter.convert(from) as T
    }

    internal class NoOpConverter : TypeConverter<Any, Any> {
        override fun convert(from: Any?): Any? {
            return from
        }
    }
}
